operator/cilium_node.go

```diff
@@ -20,16 +26,240 @@ import (
+						} else {
+							log.Warningf("Unknown CiliumNode object type %T received: %+v", newNode, newNode)
+						}
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", oldNode, oldNode)
+					}
+				},
+				DeleteFunc: func(obj interface{}) {
+					if ciliumNode := k8s.ObjToCiliumNode(obj); ciliumNode != nil {
+						deletedNode := nodeTypes.ParseCiliumNode(ciliumNode)
+						ciliumNodeKVStore.DeleteLocalKey(context.TODO(), &deletedNode)
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", obj, obj)
+					}
+				},
+			}
+		} else {
+			// nodeManager not nil thus the events will be processed by
+			// ciliumNodeKVStore and the nodeManager
+			cfg.resourceEventHandler = cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					if ciliumNode := k8s.ObjToCiliumNode(obj); ciliumNode != nil {
+						nodeNew := nodeTypes.ParseCiliumNode(ciliumNode)
+						ciliumNodeKVStore.UpdateKeySync(context.TODO(), &nodeNew)
+						// node is deep copied before it is stored in pkg/aws/eni
+						cfgBuilder.nodeManager.Create(ciliumNode)
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", obj, obj)
+					}
+				},
+				UpdateFunc: func(oldObj, newObj interface{}) {
+					if oldNode := k8s.ObjToCiliumNode(oldObj); oldNode != nil {
+						if newNode := k8s.ObjToCiliumNode(newObj); newNode != nil {
+							if oldNode.DeepEqual(newNode) {
+								return
+							}
+							nodeNew := nodeTypes.ParseCiliumNode(newNode)
+							ciliumNodeKVStore.UpdateKeySync(context.TODO(), &nodeNew)
+							// node is deep copied before it is stored in pkg/aws/eni
+							cfgBuilder.nodeManager.Update(newNode)
+						} else {
+							log.Warningf("Unknown CiliumNode object type %T received: %+v", newNode, newNode)
+						}
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", oldNode, oldNode)
+					}
+				},
+				DeleteFunc: func(obj interface{}) {
+					if ciliumNode := k8s.ObjToCiliumNode(obj); ciliumNode != nil {
+						deletedNode := nodeTypes.ParseCiliumNode(ciliumNode)
+						ciliumNodeKVStore.DeleteLocalKey(context.TODO(), &deletedNode)
+						cfgBuilder.nodeManager.Delete(ciliumNode.Name)
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", obj, obj)
+					}
+				},
+			}
+		}
+	} else {
+		if cfgBuilder.nodeManager == nil {
+			// Both nodeManager and KVStore are nil. We don't need to handle
+			// any watcher events, but we will need to keep all CiliumNodes in
+			// memory because 'ciliumNodeStore' is used across the operator
+			// to get the latest state of a CiliumNode.
+
+			// Since we won't be handling any events we don't need to convert
+			// objects.
+			cfg.ciliumNodeConvertFunc = nil
+		} else {
+			// nodeManager not nil but the KVStore is nil thus the events
+			// will only be relevant to the nodeManager.
+			cfg.resourceEventHandler = cache.ResourceEventHandlerFuncs{
```

Comment:
Do we really want 3 sets of these funcs with repeated conversion and error handling?